src/main/java/com/fauna/codec/codecs/ClassCodec.java

```diff
@@ -0,0 +1,209 @@
+            var codec = provider.get(field.getType(), ta);
+            FieldInfo info = new FieldInfo(field, name, codec);
+            fieldsList.add(info);
+            byNameMap.put(info.getName(), info);
+        }
+
+        this.shouldEscapeObject = TAGS.stream().anyMatch(byNameMap.keySet()::contains);
+        this.fields = List.copyOf(fieldsList);
+        this.fieldsByName = Map.copyOf(byNameMap);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public T decode(UTF8FaunaParser parser) {
+        try {
+            FaunaTokenType endToken = parser.getCurrentTokenType().getEndToken();
+            Object instance = createInstance();
+            setFields(instance, parser, endToken);
+            return (T) instance;
+        } catch (IOException exc) {
+            throw new ClientException(unexpectedTokenExceptionMessage(parser.getCurrentTokenType()));
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void encode(UTF8FaunaGenerator gen, T obj) throws IOException {
+        if (shouldEscapeObject) {
+            gen.writeStartEscapedObject();
+        } else {
+            gen.writeStartObject();
+        }
+        for (FieldInfo fi : fields) {
+            if (!fi.getName().startsWith("this$")) {
+                gen.writeFieldName(fi.getName());
+                try {
+                    fi.getProperty().setAccessible(true);
+                    var value = fi.getProperty().get(obj);
+                    fi.getCodec().encode(gen, value);
+                } catch (IllegalAccessException e) {
+                    throw new ClientException("Error accessing field: " + fi.getName(),
+                            e);
+                }
+            }
+        }
+        if (shouldEscapeObject) {
+            gen.writeEndEscapedObject();
+        } else {
+            gen.writeEndObject();
+        }
+    }
+
+    @Override
+    public Class<?> getCodecClass() {
+        return null;
+    }
+
+    private Object createInstance() {
+        try {
+            Class<?> clazz = Class.forName(type.getTypeName());
+            Constructor<?> constructor = clazz.getConstructor();
+            return constructor.newInstance();
+        } catch (Exception ex) {
+            throw new RuntimeException(ex);
+        }
+    }
+
+    private void setFields(Object instance, UTF8FaunaParser parser,
+                           FaunaTokenType endToken) throws IOException, IllegalAccessException {
+        String id = null;
+        String name = null;
+        Module coll = null;
+        boolean exists = true;
+        String cause = null;
+
+        while (parser.read() && parser.getCurrentTokenType() != endToken) {
+            if (parser.getCurrentTokenType() != FaunaTokenType.FIELD_NAME) {
+                throw new ClientException(unexpectedTokenExceptionMessage(parser.getCurrentTokenType()));
+            }
```

Comment:
This method is a little hard to follow, and having this many nulls at the beginning gives me the heebie jeebies, but I'm sure it works :)

Reply:
I def agree, although this is a port from the existing. I'll take a quick look to see if I can come up with a straight forward rewrite